lockbox-ios/Action/DataStoreAction.swift

```diff
@@ -0,0 +1,322 @@
+                .subscribe(onError:{ error in
+                    self.lockSubject.onError(error)
+                })
+                .disposed(by: self.disposeBag)
+    }
+
+    private func _list() {
+        self.openSubject
+                .take(1)
+                .flatMap { _ in
+                    self.checkState()
+                }
+                .flatMap { _ in
+                    self.webView.evaluateJavaScript("\(self.dataStoreName!).list()")
+                }
+                .subscribe(onError:{ error in
+                    self.listSubject.onError(error)
+                })
+                .disposed(by: self.disposeBag)
+    }
+
+    private func checkState() -> Single<Bool> {
+        return _initialized().asObservable()
+                .flatMap { initialized -> Observable<Bool> in
+                    if !initialized {
+                        throw DataStoreError.NotInitialized
+                    }
+
+                    return self._locked().asObservable()
+                }
+                .map { locked -> Bool in
+                    if locked {
+                        throw DataStoreError.Locked
+                    }
+
+                    return locked
+                }
+                .asSingle()
+    }
+}
+
+extension DataStoreActionHandler: WKScriptMessageHandler, WKNavigationDelegate {
+    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
+        self._open()
+                .take(1)
+                .subscribe()
+                .disposed(by: self.disposeBag)
+    }
+
+    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
+        let messageBody = message.body
+        switch message.name {
+        case JSCallbackFunction.OpenComplete.rawValue:
+            self.openSubject.onNext(messageBody)
+            break
+        case JSCallbackFunction.InitializeComplete.rawValue:
+            self.initializeSubject.onNext(messageBody)
+            break
+        case JSCallbackFunction.UnlockComplete.rawValue:
+            self.unlockSubject.onNext(messageBody)
+            break
+        case JSCallbackFunction.LockComplete.rawValue:
+            self.lockSubject.onNext(messageBody)
+            break
+        case JSCallbackFunction.ListComplete.rawValue:
+            guard let listBody = messageBody as? [[Any]] else {
+                self.listSubject.onError(DataStoreError.UnexpectedType)
+                break
+            }
+
+            var itemList:[Item]
+            do {
+                itemList = try listBody.map { anyList -> Item in
+                    guard let itemDictionary = anyList[1] as? [String: Any] else {
+                        throw DataStoreError.UnexpectedType
+                    }
+
+                    return try self.parser.itemFromDictionary(itemDictionary)
+                }
+            } catch {
+                self.listSubject.onError(error)
+                break
```

Comment:
you can tell I'm still a semi Swift newbie ;) 

I like the `flatMap` refactor, but I also like having the Parser throw an `InvalidDictionary` error when there's no item. I shuffled this so that the error gets sent to the dispatcher as an ErrorAction instead of rolling down the Observable chain, and all valid items are still concatenated to the list.